utils/default-helpers/src/main/scala/pl/touk/nussknacker/engine/util/functions/collection.scala

```diff
@@ -0,0 +1,278 @@
+
+  @Documentation(description = "Returns the largest element (elements must be comparable)")
+  @GenericType(typingFunction = classOf[ListElementTyping])
+  def max[T <: Comparable[T]](@ParamName("list") list: java.util.Collection[T]): T = Collections.max[T](list)
+
+  @Documentation(description = "Returns a slice of the list starting with start index (inclusive) and ending at stop index (exclusive)")
+  @GenericType(typingFunction = classOf[ListTyping])
+  def slice[T](@ParamName("list") list: java.util.Collection[T], @ParamName("start") start: Int, @ParamName("stop") stop: Int): java.util.List[T] =
+    list.asScala.slice(start, stop).toList.asJava
+
+  private val numberPriorities = Map(
+    classOf[java.lang.Short] -> 1,
+    classOf[java.lang.Byte] -> 2,
+    classOf[java.lang.Integer] -> 3,
+    classOf[java.lang.Long] -> 4,
+    classOf[java.math.BigInteger] -> 5,
+    classOf[java.lang.Float] -> 6,
+    classOf[java.lang.Double] -> 7,
+    classOf[java.math.BigDecimal] -> 8,
+  ).asInstanceOf[Map[Class[_], Int]]
+
+  private def getNumberPriority(number: Number): Int = numberPriorities
+    .getOrElse(
+      number.getClass,
+      throw new IllegalArgumentException(s"Can't calculate the sum of elements. Unexpected element of type ${number.getClass.getName}: $number")
+    )
+
+  def toBigDecimal(number: Number): java.math.BigDecimal = {
+    number match {
+      case x: java.lang.Byte => java.math.BigDecimal.valueOf(x.intValue())
+      case x: java.lang.Short => java.math.BigDecimal.valueOf(x.intValue())
+      case x: Integer => java.math.BigDecimal.valueOf(x.intValue())
+      case x: java.lang.Long => java.math.BigDecimal.valueOf(x)
+      case x: java.lang.Float => java.math.BigDecimal.valueOf(x.doubleValue())
+      case x: java.lang.Double => java.math.BigDecimal.valueOf(x)
+      case x: java.math.BigDecimal => x
+      case x: java.math.BigInteger => new java.math.BigDecimal(x)
+      case x => throw new IllegalArgumentException(s"Can't convert to BigDecimal, unsupported number of type ${x.getClass.getName}: $x")
+    }
+  }
+
+  def toBigInteger(number: Number): java.math.BigInteger = {
+    number match {
+      case x: java.lang.Byte => java.math.BigInteger.valueOf(x.intValue())
+      case x: java.lang.Short => java.math.BigInteger.valueOf(x.intValue())
+      case x: Integer => java.math.BigInteger.valueOf(x.intValue())
+      case x: java.lang.Long => java.math.BigInteger.valueOf(x)
+      case x: java.math.BigInteger => x
+      case x => throw new IllegalArgumentException(s"Can't convert to BigInteger, unsupported number of type ${x.getClass.getName}: $x")
+    }
+  }
+
+  @Documentation(description = "Returns a sum of all elements")
+  @GenericType(typingFunction = classOf[ListElementTyping])
+  def sum[T <: java.lang.Number](@ParamName("listOfNumbers") list: java.util.Collection[T]): T = {
+    if (list.isEmpty) {
+      return 0.asInstanceOf[T]
```

Comment:
Will this work if we call this method with `List[BigDecimal]`?

Reply:
I don't understand generics in Scala very well, but it seems that at runtime what the following does
```
return 0.0.asInstanceOf[T]
```
is casting `0.0` to `java.lang.Number`, no matter what the argument is.